File: plugin/sandpolis-plugin-shell/agent/src/main/java/org/s7s/plugin/shell/agent/java/Shell.java
//============================================================================//
//                                                                            //
//            Copyright © 2015 - 2022 Sandpolis Software Foundation           //
//                                                                            //
//  This source file is subject to the terms of the Mozilla Public License    //
//  version 2. You may not use this file except in compliance with the MPLv2. //
//                                                                            //
//============================================================================//
package org.s7s.plugin.shell.agent.java;

import static com.google.common.base.Preconditions.checkArgument;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.s7s.core.foundation.S7SProcess;
import org.s7s.plugin.shell.Messages.ShellCapability;

public record Shell(Path executable, Set<ShellCapability> capabilities, String version) {

	private static final Map<String, Set<ShellCapability>> knownPaths = Map.of(
			// Bash
			"/bin/bash", Set.of(ShellCapability.BASH),
			// Windows CMD
			"C:/Windows/System32/cmd.exe", Set.of(ShellCapability.CMD),
			// Powershell
			"/usr/bin/pwsh", Set.of(ShellCapability.PWSH),
			// Powershell
			"C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe", Set.of(ShellCapability.PWSH),
			// Powershell
			"C:/Windows/SysWOW64/WindowsPowerShell/v1.0/powershell.exe", Set.of(ShellCapability.PWSH),
			// ZSH
			"/usr/bin/zsh", Set.of(ShellCapability.ZSH));

	public static List<Shell> discoverShells() {
		return knownPaths.keySet().stream().map(Shell::of).collect(Collectors.toList());
	}

	public static Shell of(String executable) {
		return of(Paths.get(executable));
	}

	public static Shell of(Path executable) {
		checkArgument(Files.isExecutable(executable));

		var capabilities = new HashSet<ShellCapability>();
		String version = null;

		// Check version output
		S7SProcess.exec(executable, "--version").complete((exit, stdout, stderr) -> {
			if (exit == 0) {
				if (stdout.startsWith("zsh")) {
					capabilities.add(ShellCapability.ZSH);
					capabilities.add(ShellCapability.BASH);
					capabilities.add(ShellCapability.SH);
				}

				if (stdout.startsWith("GNU bash,")) {
					capabilities.add(ShellCapability.BASH);
					capabilities.add(ShellCapability.SH);
				}
			}
		});

		// Check help output if we need to
		S7SProcess.exec(executable, "--help").complete((exit, stdout, stderr) -> {
			if (exit == 0) {
				if (stdout.startsWith("Usage: zsh")) {
					capabilities.add(ShellCapability.ZSH);
					capabilities.add(ShellCapability.BASH);
					capabilities.add(ShellCapability.SH);
				}

				if (stdout.startsWith("GNU bash,")) {
					capabilities.add(ShellCapability.BASH);
					capabilities.add(ShellCapability.SH);
				}
			}
		});

		return new Shell(executable, Collections.unmodifiableSet(capabilities), version);
	}

	public ProcessBuilder execute(String command) {
		if (capabilities.contains(ShellCapability.SH)) {
			return new ProcessBuilder(executable.toString(), "-c",
					"echo " + Base64.getEncoder().encodeToString(command.getBytes()) + " | base64 --decode | "
							+ executable.toString());
		}

		if (capabilities.contains(ShellCapability.CMD)) {
			return new ProcessBuilder(executable.toString(), "/C", command);
		}

		if (capabilities.contains(ShellCapability.PWSH)) {
			return new ProcessBuilder(executable.toString(), "-encodedCommand",
					Base64.getEncoder().encodeToString(command.getBytes(StandardCharsets.UTF_16LE)));
		}

		throw new UnsupportedOperationException();
	}

	public ProcessBuilder newSession() {
		if (capabilities.contains(ShellCapability.SH)) {
			return new ProcessBuilder(executable.toString(), "-i");
		}

		return new ProcessBuilder(executable.toString());
	}
}

File: sandpolis-shell/src/shell.rs
use anyhow::{Result, anyhow};
use base64::{Engine as _, engine::general_purpose};
use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};
use std::process::Command;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum ShellCapability {
    Bash,
    Cmd,
    Pwsh,
    Sh,
    Zsh,
}

#[derive(Debug, Clone)]
pub struct Shell {
    pub executable: PathBuf,
    pub capabilities: HashSet<ShellCapability>,
    pub version: Option<String>,
}

impl Shell {
    fn known_paths() -> HashMap<&'static str, HashSet<ShellCapability>> {
        let mut paths = HashMap::new();

        // Bash
        paths.insert("/bin/bash", [ShellCapability::Bash].into());

        // Windows CMD
        paths.insert("C:/Windows/System32/cmd.exe", [ShellCapability::Cmd].into());

        // PowerShell
        paths.insert("/usr/bin/pwsh", [ShellCapability::Pwsh].into());
        paths.insert(
            "C:/Windows/System32/WindowsPowerShell/v1.0/powershell.exe",
            [ShellCapability::Pwsh].into(),
        );
        paths.insert(
            "C:/Windows/SysWOW64/WindowsPowerShell/v1.0/powershell.exe",
            [ShellCapability::Pwsh].into(),
        );

        // ZSH
        paths.insert("/usr/bin/zsh", [ShellCapability::Zsh].into());

        paths
    }

    pub fn discover_shells() -> Vec<Shell> {
        Self::known_paths()
            .keys()
            .filter_map(|path| Self::of_path(PathBuf::from(path)).ok())
            .collect()
    }

    pub fn of(executable: &str) -> Result<Shell> {
        Self::of_path(PathBuf::from(executable))
    }

    pub fn of_path(executable: PathBuf) -> Result<Shell> {
        if !executable.exists() || !Self::is_executable(&executable) {
            return Err(anyhow!("File is not executable: {:?}", executable));
        }

        let mut capabilities = HashSet::new();
        let mut version = None;

        // Check version output
        if let Ok(output) = Command::new(&executable).arg("--version").output()
            && output.status.success() {
                let stdout = String::from_utf8_lossy(&output.stdout);

                if stdout.starts_with("zsh") {
                    capabilities.insert(ShellCapability::Zsh);
                    capabilities.insert(ShellCapability::Bash);
                    capabilities.insert(ShellCapability::Sh);
                }

                if stdout.starts_with("GNU bash,") {
                    capabilities.insert(ShellCapability::Bash);
                    capabilities.insert(ShellCapability::Sh);
                }

                // Extract version if possible
                version = Some(stdout.lines().next().unwrap_or("").to_string());
            }

        // Check help output if we need more information
        if capabilities.is_empty()
            && let Ok(output) = Command::new(&executable).arg("--help").output()
                && output.status.success() {
                    let stdout = String::from_utf8_lossy(&output.stdout);

                    if stdout.starts_with("Usage: zsh") {
                        capabilities.insert(ShellCapability::Zsh);
                        capabilities.insert(ShellCapability::Bash);
                        capabilities.insert(ShellCapability::Sh);
                    }

                    if stdout.starts_with("GNU bash,") {
                        capabilities.insert(ShellCapability::Bash);
                        capabilities.insert(ShellCapability::Sh);
                    }
                }

        // If still no capabilities found, try to infer from known paths
        if capabilities.is_empty()
            && let Some(known_caps) = Self::known_paths().get(executable.to_str().unwrap_or("")) {
                capabilities = known_caps.clone();
            }

        Ok(Shell {
            executable,
            capabilities,
            version,
        })
    }

    pub fn execute(&self, command: &str) -> Result<Command> {
        let mut cmd = Command::new(&self.executable);

        if self.capabilities.contains(&ShellCapability::Sh) {
            // Use base64 encoding for POSIX shells to handle special characters
            let encoded_command = general_purpose::STANDARD.encode(command.as_bytes());
            cmd.arg("-c").arg(format!(
                "echo {} | base64 --decode | {}",
                encoded_command,
                self.executable.display()
            ));
        } else if self.capabilities.contains(&ShellCapability::Cmd) {
            cmd.arg("/C").arg(command);
        } else if self.capabilities.contains(&ShellCapability::Pwsh) {
            // PowerShell expects UTF-16LE for encoded commands
            let utf16_bytes: Vec<u8> = command
                .encode_utf16()
                .flat_map(|u| u.to_le_bytes().to_vec())
                .collect();
            let encoded_command = general_purpose::STANDARD.encode(&utf16_bytes);
            cmd.arg("-encodedCommand").arg(encoded_command);
        } else {
            return Err(anyhow!("Unsupported shell capabilities"));
        }

        Ok(cmd)
    }

    pub fn new_session(&self) -> Command {
        let mut cmd = Command::new(&self.executable);

        if self.capabilities.contains(&ShellCapability::Sh) {
            cmd.arg("-i");
        }

        cmd
    }

    fn is_executable(path: &Path) -> bool {
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            if let Ok(metadata) = path.metadata() {
                let permissions = metadata.permissions();
                permissions.mode() & 0o111 != 0
            } else {
                false
            }
        }

        #[cfg(windows)]
        {
            // On Windows, check if it's a .exe, .cmd, .bat file or if it exists
            path.extension()
                .and_then(|ext| ext.to_str())
                .map(|ext| matches!(ext.to_lowercase().as_str(), "exe" | "cmd" | "bat" | "ps1"))
                .unwrap_or(false)
                || path.exists()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_known_paths() {
        let paths = Shell::known_paths();
        assert!(!paths.is_empty());
        assert!(paths.contains_key("/bin/bash"));
    }

    #[test]
    fn test_shell_creation_with_nonexistent_path() {
        let result = Shell::of("/nonexistent/path");
        assert!(result.is_err());
    }
}
